File: launcher/minimal/src/main/java/ink/reactor/launcher/logger/type/FileLogger.java
package ink.reactor.launcher.logger.type;

import ink.reactor.launcher.logger.data.LoggerLevels;
import ink.reactor.launcher.logger.file.FileWriter;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
public final class FileLogger {
    private final LoggerLevels loggerLevels;
    private final FileWriter fileWriter;

    public void debug(final String message) {
        if (loggerLevels.debug()) {
            fileWriter.write(message.getBytes());
        }
    }

    public void log(final String message) {
        if (loggerLevels.log()) {
            fileWriter.write(message.getBytes());
        }
    }

    public void info(final String message) {
        if (loggerLevels.info()) {
            fileWriter.write(message.getBytes());
        }
    }

    public void warn(final String message) {
        if (loggerLevels.warn()) {
            fileWriter.write(message.getBytes());
        }
    }

    public void error(final String message) {
        if (loggerLevels.error()) {
            fileWriter.write(message.getBytes());
        }
    }

    public void error(final String message, final Throwable throwable) {
        if (loggerLevels.error()) {
            fileWriter.write(message.getBytes());
            fileWriter.write(('\n' + throwable.toString()).getBytes());

            for (StackTraceElement element : throwable.getStackTrace()) {
                fileWriter.write(("\n\tat " + element.toString()).getBytes());
            }

            Throwable cause = throwable.getCause();
            while (cause != null) {
                fileWriter.write(("\nCaused by: " + cause).getBytes());
                for (StackTraceElement element : cause.getStackTrace()) {
                    fileWriter.write(("\n\tat " + element.toString()).getBytes());
                }
                cause = cause.getCause();
            }
        }
    }
}

File: launcher/minimal/src/main/kotlin/ink/reactor/launcher/logger/type/FileLogger.kt
package ink.reactor.launcher.logger.type

import ink.reactor.launcher.logger.LoggerLevels
import ink.reactor.launcher.logger.file.FileWriter

class FileLogger(
    private val loggerLevels: LoggerLevels,
    private val fileWriter: FileWriter
) {

    fun debug(message: String) {
        if (loggerLevels.debug) {
            fileWriter.write(message.toByteArray())
        }
    }

    fun log(message: String) {
        if (loggerLevels.log) {
            fileWriter.write(message.toByteArray())
        }
    }

    fun info(message: String) {
        if (loggerLevels.info) {
            fileWriter.write(message.toByteArray())
        }
    }

    fun warn(message: String) {
        if (loggerLevels.warn) {
            fileWriter.write(message.toByteArray())
        }
    }

    fun error(message: String) {
        if (loggerLevels.error) {
            fileWriter.write(message.toByteArray())
        }
    }

    fun error(message: String, throwable: Throwable) {
        if (loggerLevels.error) {
            fileWriter.write(message.toByteArray())
            fileWriter.write(('\n'.toString() + throwable.toString()).toByteArray())

            for (element in throwable.stackTrace) {
                fileWriter.write(("\n\tat $element").toByteArray())
            }

            var cause = throwable.cause
            while (cause != null) {
                fileWriter.write(("\nCaused by: $cause").toByteArray())
                for (element in cause.stackTrace) {
                    fileWriter.write(("\n\tat $element").toByteArray())
                }
                cause = cause.cause
            }
        }
    }
}
